src/proposals/20230802/SparkGoerli_20230802.t.sol

```diff
@@ -0,0 +1,390 @@
+        this._borrow(weth, POOL, user2, 350e18, false);
+
+        _deposit(dai,  POOL, user3, 1_000_000e18);
+        _deposit(weth, POOL, user3, 1_000e18);
+        this._borrow(weth, POOL, user3, 1_000e18, false);
+
+        _executePayload(address(payload));
+
+        // --- Test 1 - Cannot borrow more than small amount against DAI ---
+
+        // Deposit 1m
+        _deposit(dai, POOL, user1, 1_000_000e18);
+
+        uint256 wethPrice = ORACLE.getAssetPrice(WETH);
+
+        assertEq(wethPrice, 1_869.21537501e8);
+
+        // 1m * 0.01% => $100, divided by WETH price to get max borrow
+        uint256 maxWethBorrow = 100e18 * 1e8 / wethPrice;
+
+        assertApproxEqAbs(maxWethBorrow, 0.054e18, 0.002e18);
+
+        // Should only be able to borrow small amounts of ETH
+        this._borrow(weth, POOL, user1, maxWethBorrow - 0.00001e18, false);
+
+        // Cannot borrow more
+        vm.expectRevert(bytes('36'));	// COLLATERAL_CANNOT_COVER_NEW_BORROW
+        this._borrow(weth, POOL, user1, 0.00001e18, false);
+
+        // --- Test 2 - Can liquidate any single position that was previously setup ---
+
+        // Liquidate the position setup previously
+        assertEq(IERC20(dai.underlying).balanceOf(liquidator1), 0);
+        assertEq(IERC20(dai.aToken).balanceOf(user2),           1_000_000e18);
+
+        _liquidate(dai, weth, POOL, liquidator1, user2, 350e18);
+
+        // Liquidator should get about 650,000 DAI (350 ETH with price at ~$1,870 / ETH)
+        assertApproxEqAbs(IERC20(dai.underlying).balanceOf(liquidator1), 650_000e18, 30_000e18);
+
+        // User can keep remainder
+        assertApproxEqAbs(IERC20(dai.aToken).balanceOf(user2), 350_000e18, 40_000e18);
+
+        // --- Test 3 - Liquidate multi-collateralized position ---
+
+        // We can fully liquidate the DAI position which now contributes almost nothing to HF
+        assertEq(IERC20(dai.underlying).balanceOf(liquidator2),  0);
+        assertEq(IERC20(weth.underlying).balanceOf(liquidator2), 0);
+        assertEq(IERC20(dai.aToken).balanceOf(user3),            1_000_000e18);
+        assertEq(IERC20(weth.aToken).balanceOf(user3),           1_000e18);
+
+        // Can only liquidate about half the debt, but this will make the position healthy
+        // Can only do half because there is only 1m DAI collateral for 2m in debt
+        _liquidate(dai, weth, POOL, liquidator2, user3, 1_000e18);
+
+        assertApproxEqAbs(IERC20(dai.underlying).balanceOf(liquidator2), 1_000_000e18, 50_000e18);
+
+        // Some WETH is leftover because the liquidation call was limited by the amount of DAI available
+        assertApproxEqAbs(IERC20(weth.underlying).balanceOf(liquidator2), 500e18, 50e18);
+        assertApproxEqAbs(IERC20(dai.aToken).balanceOf(user3),            0,      1);
+    }
+
+    function _getAnnualizedDsr(uint256 dsr) internal pure returns (uint256) {
+        return (dsr - RAY) * 365 days;
+    }
```

Comment:
How does this expression give the annualized DSR?
Isn't the DSR defined as a per-second rate?

Reply:
Yeah the `dsr` as the argument is the per-second rate which looks like `100.000000123%`. So subtracting a `RAY` (100%) will give you a value `0.00000000123%` and then multiplying by the seconds in the year gives you the APR (not APY).

Reply:
Yeah the `dsr - RAY` value gives the yield per second, then I'm multiplying that by one year in seconds (`365 days`).